minio/minio
# Use hdfs.Readdir() to optimize HDFS directory listings

## Description

Currently, listing directories on HDFS incurs a per-entry remote Stat() call penalty, the cost of which can really blow up on large directories with many entries (+1,000) especially when considered in addition to peripheral calls (such as validation) and the fact that minio is an intermediary to the client (whereas other clients listed below can query HDFS directly).

Because listing directories this way is expensive, the Golang HDFS library provides the [`Client.Open()`] function which creates a [`FileReader`] that is able to batch multiple calls together through the [`Readdir()`] function.

This is substantially more efficient for very large directories.

## Motivation and Context

In one case we were witnessing about +20 seconds to list a directory with 1,500 entries, admittedly large, but the Java hdfs ls utility as well as the Golang HDFS library sample ls utility were much faster.

Hadoop HDFS DFS (4.02s):

    λ ~/code/minio → use-readdir
    » time hdfs dfs -ls /directory/with/1500/entries/
    …
    hdfs dfs -ls   5.81s user 0.49s system 156% cpu 4.020 total

Golang HDFS library (0.47s):

    λ ~/code/hdfs → master
    » time ./hdfs ls -lh /directory/with/1500/entries/
    …
    ./hdfs ls -lh   0.13s user 0.14s system 56% cpu 0.478 total

mc and minio **without** optimization (16.96s):

    λ ~/code/minio → master
    » time mc ls myhdfs/directory/with/1500/entries/
    …
    ./mc ls   0.22s user 0.29s system 3% cpu 16.968 total

mc and minio **with** optimization (0.40s):

    λ ~/code/minio → use-readdir
    » time mc ls myhdfs/directory/with/1500/entries/
    …
    ./mc ls   0.13s user 0.28s system 102% cpu 0.403 total

I also received a report that on a directory of +20,000 entries, it reduced `ls` times from 4 minutes down to 1.27 seconds.

[`Client.Open()`]: https://godoc.org/github.com/colinmarc/hdfs#Client.Open
[`FileReader`]: https://godoc.org/github.com/colinmarc/hdfs#FileReader
[`Readdir()`]: https://godoc.org/github.com/colinmarc/hdfs#FileReader.Readdir

## How to test this PR?

List a large HDFS directory (mine had 1,500 entries).

You can probably create such a directory like this. Note that I include a sub-directory to test mixed contents.

``` bash
$ TEST_PATH=/some/path
$ hdfs dfs -mkdir -p $TEST_PATH/test_sub_dir
$ hdfs dfs -touchz $TEST_PATH/test_sub_dir/test_child_file
$ for i in {0001..1500}; do hdfs dfs -touchz "$TEST_PATH/file_${i}"; done
```

That might be slow since it's invoking `hdfs` per file, so you can probably batch them into one call like this:

``` bash
$ TEST_PATH=/some/path
$ hdfs dfs -mkdir -p $TEST_PATH/test_sub_dir
$ hdfs dfs -touchz $TEST_PATH/test_sub_dir/test_child_file
$ for i in {0001..1500}; do echo "$TEST_PATH/file_${i}"; done | xargs hdfs dfs -touchz
```

Then you can test by listing it:

``` bash
$ ./mc ls myhdfs$TEST_PATH
$ ./mc ls --recursive myhdfs$TEST_PATH
```

## Types of changes

## Checklist:
